Added `box-sizing: border-box;` to...

joshdarby opened the pull request:
`figcaption` and `.wp-caption-text` elements so that the caption text doesn't overflow from the parent container. Issue #1702

Commit:
Added `box-sizing: border-box;` attr to figcaptions and wp-caption-text

Commit:
Merge branch '0.5-dev' into 1702-add-box-sizing-attribute-to-gallery-captions

Commit:
Added `box-sizing` attr to one more `figcaption`

Commit:
Updated css files for `figcaption` that had `box-sizing` attr added

Commit:
Updated changelog for PR #1711

joshdarby commented:
@benlk Can you review this and let me know if you think we need to add the `box-sizing` attribute to any other versions of `figcaption` and `.wp-caption-text`?

benlk committed:
Merge branch '0.5-dev' into 1702-add-box-sizing-attribute-to-gallery-captions

benlk commented:
`.wp-block-gallery`, which I'll add some styles for in here.

benlk commented on less/inc/images.less:
```diff
@@ -28,6 +28,9 @@
         -o-object-fit: scale-down;
         object-fit: scale-down;
       }
+      figcaption {
```
This selector is `.wp-block-gallery.is-cropped .blocks-gallery-item figcaption` but not all galleries are `.is-cropped`.

benlk committed:
Remove .is-cropped selector for box-sizing on figcaption in .wp-block-gallery; other changes:

- make the figcaption no more than 50% of height of container
- left align text to match other captions
- reduce padding on caption

benlk commented:
The padding/max-height changes in https://github.com/INN/largo/pull/1711/commits/df15461147fb02201cdacef5afd607098b593da9 are to make sure that with Very Long captions, the caption doesn't fully cover the image.

![Screen Shot 2019-06-17 at 19 14 05 ](https://user-images.githubusercontent.com/1754187/59642696-19483400-9134-11e9-85e2-992bbfa682d5.png)
![Screen Shot 2019-06-17 at 19 14 24 ](https://user-images.githubusercontent.com/1754187/59642706-249b5f80-9134-11e9-898f-d9f3e424a38e.png)

But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?

![Screen Shot 2019-06-17 at 20 00 50 ](https://user-images.githubusercontent.com/1754187/59644355-a098a600-913a-11e9-8336-314dee24f95c.png)


If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?

Classic image:
![Screen Shot 2019-06-17 at 19 59 11 ](https://user-images.githubusercontent.com/1754187/59644338-88288b80-913a-11e9-9f1c-fd60b59eaffb.png)

Gallery block:
![Screen Shot 2019-06-17 at 19 59 16 ](https://user-images.githubusercontent.com/1754187/59644349-924a8a00-913a-11e9-9978-845aab7593de.png)

joshdarby commented:
> But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?

@benlk I'm not opposed to either way. I'd think if having them below matches Largo styles more, we should go that route.

> If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?

If 3/4 of them are showing the caption in the expanded view, I'd think the 4th should show it also. Unless you think there's a use case where someone using the theme would want one instance to use that doesn't show the captions in that view?

benlk commented:
> > If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?
> 
> If 3/4 of them are showing the caption in the expanded view, I'd think the 4th should show it also. Unless you think there's a use case where someone using the theme would want one instance to use that doesn't show the captions in that view?

I don't believe there is such a case; let's make those cases consistent. 

>> But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?
> 
> @benlk I'm not opposed to either way. I'd think if having them below matches Largo styles more, we should go that route.

My main worry here is that, since we have Gutenberg's default stylesheet enqueued, Gutenberg will change something upstream will horribly conflict with the styles we apply here, and as a result our sites will complain.

I'd love for there to be a way to get Gutenberg to enqueue its styles per-block, so that we would be able to dequeue the Gallery Block styles and replace those, but then let Gutenberg enqueue the styles for New Shiny Block until we can write custom styles for it.

Should we go ahead and make drastic changes to CSS without dequeueing Gutenberg's styles?

benlk commented:
> Should we go ahead and make drastic changes to CSS without dequeueing Gutenberg's styles?

No.

Therefore, let's not change the gallery block caption display location, yet.

benlk commented:
Next steps:

- [x] Make sure all captions display consistently in slideshow view.

joshdarby commented:
> * [ ]  Make sure all captions display consistently in slideshow view.

@benlk Should they all display the way they do in the classic image or in the Gutenberg block?

benlk commented:
My personal preference is the classic image view, because that's easier to read than the gradient background.

Commit:
Merge branch '0.5-dev' into 1702-add-box-sizing-attribute-to-gallery-captions

Commit:
Updated navis slideshow slides to have transparent background,

also updated single image block expanded view caption styling to match the classic image.

joshdarby commented:
@benlk Now the captions throughout classic image, image gallery block, and classic gallery shortcode are consistent. The only inconsistency is the known issue of Gutenberg image blocks/galleries not displaying the media credits from issue #1683.

benlk commented:
Sounds good; can you shove screenshots in this ticket and update the changelog before we merge?

Commit:
Updated changelog.md

joshdarby commented:
@benlk 

Classic image block:
![Screen Shot 2019-06-24 at 1 42 30 PM](https://user-images.githubusercontent.com/18353636/60040136-48a3f700-9686-11e9-8627-5e8bd8481bc6.png)

Classic image gallery:
![Screen Shot 2019-06-24 at 1 42 13 PM](https://user-images.githubusercontent.com/18353636/60040139-48a3f700-9686-11e9-99d9-bda43e0db53c.png)

Gutenberg gallery block:
![Screen Shot 2019-06-24 at 1 42 23 PM](https://user-images.githubusercontent.com/18353636/60040137-48a3f700-9686-11e9-9379-d2c3f5adc1a7.png)

benlk commented on less/inc/images.less:
```diff
@@ -30,4 +30,19 @@
+      figcaption {
+        background-color: transparent;
+        color: #ccc;
+        padding: 1em 17.023%!important;
```
It matches https://github.com/INN/largo/blob/c41e28e6381397be1918184b906ec84bc019745b/lib/navis-slideshows/css/slides.css#L310-L312 , but yes, this should be documented in code.

benlk committed:
Add comment explaining where spacing comes from.

benlk committed:
Move block styles for navis to blocks.less

benlk committed:
Remove style hiding figcaption on image block in slideshow.

benlk committed:
Note for future about media credit positioning

benlk commented:
The background shouldn't be flat transparent; it should be `rgba(0,0,0,.7);` to match the Navis slideshow styles:

https://github.com/INN/largo/blob/c41e28e6381397be1918184b906ec84bc019745b/lib/navis-slideshows/css/slides.css#L308

Also, the figcaption in wp-block-image in the Navis slideshow state doesn't match these styles; the caption is hidden.

-----

**Edit**: this is fixed now.

benlk commented:
When testing images, we need to test all of these separate sets of conditions:

- caption state
    1. no caption
    2. caption
    3. media credit
        - name
        - name and org
        - name and org and link
    4. caption and media credit
- image type
    1. classic editor image
    2. classic editor gallery shortcode
    3. block editor image
    4. block editor gallery
- alignment options (block and classic)
    1. unset
    2. left
    3. center
    4. right
    5. wide
    6. full

That's ... 6 * 4 * 6 = 144 different potential image types to test. 288 if we include the clicked-on Navis Slideshow as another variable.

We need a plan for testing this stuff systematically, or at least a stock set of markup to test against that generates all those options.

benlk commented:
Also, working on this changeset and similar ones has made me doubt the usefulness of splitting the Navis Slideshow CSS out from other image-related CSS in Largo; there's a lot of common variables. I'd be in favor of generating it using the Grunt workflow from LESS, where we could set shared variables between the other image code and the slideshow code. The slideshow CSS/JS should still only be enqueued when needed.

benlk committed:
Fix caption positioning on aligned image blocks and image blocks

benlk commented:
Another image case to consider:

- WordPress gallery widget
- Largo image widget
- WordPress image widget

joshdarby commented:
@benlk Do you want to get together and figure out a plan for acceptable criteria for getting this PR merged?

benlk commented:
Yeah, let's set aside time tomorrow to test this.

benlk commented:
IIRC, the plan we had discussed on that day was:

- test against the test content we have available to us now
- postpone the Grand Generation of Test Content until a later release

Commit:
Merge branch '0.5-dev'

joshdarby commented:
@benlk I created a sample post [here](http://largodev.innsandbox.wpengine.com/2019/07/17/box-sizing-examples/) with a 

- [ ] classic block image 
- [ ] classic block gallery
- [ ] image block
- [ ] gallery block

What else do you think we should test for this?

Commit:
Merge branch '0.5-dev' into 1702-add-box-sizing-attribute-to-gallery-captions

benlk commented:
Do we know what's going on in Firefox where opening the modal scrolls the browser to the top of the page, triggering the appearance of the sticky nav? https://github.com/INN/largo/issues/1745

![Screen Shot 2019-07-17 at 16 14 10 ](https://user-images.githubusercontent.com/1754187/61408434-013c0f80-a8ae-11e9-8616-9883a4e2de82.png)

benlk commented:
There's a difference in how the image block and Gallery block display their images, with whether the images are stretched to full size:

```html
<figure class="wp-block-image navis-slideshow navis-single navis-full" style="max-width: 100%;">
	<span class="navis-before">X</span>
	<img
		src="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png"
		alt=""
		class="wp-image-1884"
		srcset="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png 420w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-140x140.png 140w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-336x333.png 336w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-60x60.png 60w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-96x96.png 96w"
	>
	<figcaption>Largo is a responsive WordPress theme framework designed for news publishers. Finely-crafted and expertly-honed by the product and technology team at the Institute for Nonprofit News.</figcaption>
</figure>

```

![Screen Shot 2019-07-17 at 16 16 52 ](https://user-images.githubusercontent.com/1754187/61408611-6bed4b00-a8ae-11e9-9b29-f3a1dafea5db.png)


```html
<ul class="wp-block-gallery columns-1 is-cropped">
	<li class="blocks-gallery-item">
		<figure class="navis-slideshow navis-single navis-full" style="max-width: 100%;">
			<span class="navis-before">X</span>
			<img
				src="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png"
				alt=""
				data-id="1884"
				data-link="http://largodev.innsandbox.wpengine.com/?attachment_id=1884"
				class="wp-image-1884"
				srcset="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png 420w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-140x140.png 140w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-336x333.png 336w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-60x60.png 60w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-96x96.png 96w"
			>
			<figcaption>Largo is a responsive WordPress theme framework designed for news publishers. Finely-crafted and expertly-honed by the product and technology team at the Institute for Nonprofit News.</figcaption>
		</figure>
	</li>
</ul>
```
![Screen Shot 2019-07-17 at 16 16 57 ](https://user-images.githubusercontent.com/1754187/61408603-66900080-a8ae-11e9-9d9a-9a1182abbc32.png)

joshdarby commented:
> There's a difference in how the image block and Gallery block display their images, with whether the images are stretched to full size:

@benlk I think that's a known issue, right? And is it in the scope of #1702 to fix in this PR?

benlk commented:
I'm not sure it's a known issue; I'll create a ticket for it and let's merge this. 

https://github.com/INN/largo/issues/1745 isn't in this milestone either.
